pkg/endpoint/endpoint.go

```diff
@@ -2280,3 +2281,112 @@ func (e *Endpoint) SyncEndpointHeaderFile() error {
 	e.dnsHistoryTrigger.Trigger()
 	return nil
 }
+
+type ipReleaser interface {
+	ReleaseIP(net.IP) error
+}
+
+type monitorOwner interface {
+	NotifyMonitorDeleted(e *Endpoint)
+}
+
+type EndpointRemover interface {
+	Remove(ep *Endpoint) <-chan struct{}
+}
+
+// Delete cleans up all resources associated with this endpoint, including the
+// following:
+// * all goroutines managed by this Endpoint (EventQueue, Controllers)
+// * removal from the endpointmanager, resulting in new events not taking effect
+// on this endpoint
+// * cleanup of datapath state (BPF maps, proxy configuration, directories)
+// * releasing IP addresses allocated for the endpoint
+// * releasing of the reference to its allocated security identity
+func (e *Endpoint) Delete(monitor monitorOwner, ipam ipReleaser, manager EndpointRemover, conf DeleteConfig) []error {
```

Comment:
since we still have the `endpointmanager` as a global variable this function could accept a `func(ep *endpoint.Endpoint) <-chan struct{}` instead of a `EndpointRemover`.

And then in [0] one could call `ep.Delete(emd, d.ipam, endpointmanager.Remove, conf)`

Reply:
@aanm I am going to wait for the endpointmanager refactor to merge before I mark this as ready for review.